CRM/MembersOnlyEvent/BAO/EventGroup.php

```diff
@@ -130,29 +128,66 @@ public static function getContactGroupIDs($contactID) {
     if (self::$contactGroupIDs !== NULL) {
       return self::$contactGroupIDs;
     }
-    self::$contactGroupIDs = [];
 
+    self::$contactGroupIDs = array_merge(
+      self::getContactNormalGroupIds($contactID),
+      self::getContactSmartGroupIds($contactID)
+    );
+
+    return self::$contactGroupIDs;
+  }
+
+  /**
+   * Gets the normal groups for the specified contact.
+   *
+   * @param int $contactID
+   *
+   * @return array
+   *   List of group ids or empty array if nothing found
+   */
+  public static function getContactNormalGroupIds($contactID) {
     $params = [
       'sequential' => 1,
       'contact_id' => (int) $contactID,
       'status' => 'Added',
       'options' => ['limit' => 0],
-      // The return param is not working.
-      /* 'return' => ['group_id'], */
-      // The IN param is not working.
-      /* 'group_id' => ['IN' => $allowedGroups], */
     ];
 
     $contactGroups = civicrm_api3('GroupContact', 'get', $params);
 
+    $groupIDs = [];
     if ($contactGroups['count']) {
       $contactGroupIDs = array_map(function($contactGroup) {
         return $contactGroup['group_id'];
       }, $contactGroups['values']);
-      self::$contactGroupIDs = $contactGroupIDs;
+      $groupIDs = $contactGroupIDs;
     }
 
-    return self::$contactGroupIDs;
+    return $groupIDs;
+  }
+
+  /**
+   * Gets the smarty groups for the specified contact.
+   *
+   * @param int $contactID
+   *
+   * @return array
+   *   List of group ids or empty array if nothing found
+   */
+  public static function getContactSmartGroupIds($contactID) {
+    // Gets the smart group ids that the contact belongs to.
```

Comment:
no need for this inline comment, the docblock already explains what this method is about.